RFC-0017-PyTorch-Operator-Versioning.md

```diff
@@ -0,0 +1,223 @@
+
+
+## Goals
+
+
+
+* To support backward and (some) forward compatibility for an arbitrary BC or FC break update (schema, functional, etc) on an operator by [versioning](https://docs.google.com/document/d/1nyXmss2O003ZgKrhDmd-kyLNjjMqEXww_2skOXqkks4/edit). 
+* It’s a systematic flow to prevent BC/FC breakage on both deploy and runtime stages.
+* To provide testing that accurately detects dangerous BC and FC-breaking changes.
+* To support both server and edge use cases, including TorchScript, package/deploy and Edge.
+
+
+## Non-goals
+
+
+
+* It does not mean that models with old operator schema can **always** run successfully on new runtime and vice versa. 
+* It’s not for the “automatic” BC/FC support that can be done without any developer’s manual work (for example, the number of arguments mentioned in the Context). To apply versioning on the updated operator, the author needs to manually add a line in the version table and provide the resolution function for BC. This proposal is for BC/FC breakages that the automatic supports don’t apply. 
+* It does not include the BC/FC for package/deploy itself. The Python-only operators are transparent to TS and Edge clients, with the TS compilation. 
+
+
+# Proposal
+
+
+
+* **Eager changes**
+    * operator_versions.yaml** and **operator_upgraders.py** are added to register operator upgrades that are BC/FC breaking. **
+
+<p id="gdcalert1" ><span style="color: red; font-weight: bold">>>>>>  gd2md-html alert: undefined internal link (link text: "Example changes"). Did you generate a TOC? </span><br>(<a href="#">Back to top</a>)(<a href="#gdcalert2">Next alert</a>)<br><span style="color: red; font-weight: bold">>>>>> </span></p>
+
+[Example changes](#heading=h.enajdjh00ukn)
+        * Note: this will not cover functional operators
+        * The default value is zero
+    * **Use a single operator version number for all operators**
+        * This number may be shared by the deploy version, but separate from other file format versions
+    * **Older versions of operators registration are kept but they should only be matchable with special circumstances**
+    * **Newer version of the operator registry must specify an upgrader** that conforms to the older version of the operator schema. Its body is a TorchScript-able function that uses the newer version of operator to implement old semantics.
+    * [Improved BC testing] Tests that the old serialized version of the operator can still be loaded on the new runtime and run as expected need to be easy to add
+        * This seems straightforward for Torchscript and Edge testing but I’m not sure how it would work for deploy/package
```

Comment:
deploy/package should adhere to a deprecation window and not use upgraders

in the future we expect torchscript and mobile to also use the deprecation window and not use upgraders

Reply:
> deploy/package should adhere to a deprecation window and not use upgraders
> 
> in the future we expect torchscript and mobile to also use the deprecation window and not use upgraders

Reply:
No, in the deprecation window the behavior throws a warning but PyTorch will still execute it

Reply:
Oh, to my understanding the window can be quite big. Not sure if it's a good solution for mobile. If it's future plan, I'll put it in the open question session and put more data there.

Reply:
I propose a 90 day deprecation window